### PR TITLE
Normalize type_expr nodes on access

### PR DESCRIPTION
**Pull request by garrigue**
This PR was written in collaboration with @t6s .

This is a follow-up to the introduction of `private type_expr`.
We go further by making `type_expr` abstract, having the accessor functions (`get_desc`,`get_level`,...) systematically call `repr` before returning their result. This provides a clearer semantics for `type_expr`, independent of the physical  representation, which should simplify reasoning for everybody. In the future, this could also allow to replace the current in-place union-find by another mechanism.
This greatly increases the number of calls to `repr` (from 1.7x for coqtop to 3.7x for stdlib), but this also ensures that we never forget it. The performance measurements show a 5-10% overhead for `ocamlc`, but this goes down to 2-5% for `ocamlc.opt` (with one level of unrolling) (2% for coqc, 4% for lablgtk or ocamldoc).

The private `type_expr` is kept as `transient_expr`, with operations in module `Types.Transient_expr` and is used in a limited number of places. It is actually smaller than expected; the only place where it seems really necessary is to partly undo unification in `unify`. There may be a better way to provide error messages. Otherwise, this is mostly to be able to use `transient_expr` as an ordered type, for functors. As an exception, `Printtyp` uses extensively `transient_type`, as its role is to print the graph, using many tables; we made less efforts to avoid it, as correctness matters less there. 

This PR contains 4 commits:
* The first one makes `type_expr` abstract, introduces `transient_expr`, and uses `repr : type_expr -> transient_expr` for normalization. `repr` and logging/backtracking also move to `types.ml`. It shows that tracking normalization through types is useful, but painful whereas it is not really safe (a `transient_expr` can still be turned into a `Tlink` by a hidden side-effect), hence our choice to try the `always_repr` approach.
* The second one hides `repr` inside the `Transient_expr` module, and replaces almost all uses of `repr` by accessor functions. This creates an extensive abstraction barrier around `type_expr`.
* The third one fixes two bugs uncovered by the transition. One is that `expand_trace` keeps old expansions, which may create an incoherent trace if the type was modified (see the error message in `pr4018_bad.compilers.reference`). The other is that `copy_sep` should copy the proxy (end of the row) if it copies one of the fields (see `poly.ml`).
* The last one unrolls `repr_link` once to gain a few percents of performance.

In the process we also found a few places where `repr` was missing, in `Typedecl` for instance.

Some possible improvements are not done in this PR, and left for later:
* use a GADT to remove `Tlink` from the result of `get_desc`,
* replace `try_mark_node` by `try_logged_mark_node` everywhere.

### CODE REVIEW

**Comment by garrigue** [comment]
No reaction yet...
We hope that it will relieve a large part of the stress of writing correct typing code.

Concerning the overhead, we are currently collecting data on what changed, and why the impact depends much on the codebased. We will post the results when we are ready.

**Comment by Octachron** [comment]
My early reaction is that this looks really promising in term of readability of the code; even if the performance change is a bit worrying.

**Comment by gasche** [comment]
Another early reaction:
- Happy to see work in this direction.
- I'm fine with the performance drop. I think that many changes that affect type representation may have constant-factor impact on the type-checking performance, and we should prioritize maintainability and readability over those constant factors; as far as I know, the large type-checking times reported on eg. the Jane Street codebase are not due to constant-factor issues, but to inefficiencies in the algorithms, especially at the level of modules. (cc @lpw25 does that sound reasonable?)
- I have a small review budget right now and I prioritize #10219 (then Thomas is busy with #8516 and I'm sure Octachron also has a big PR in hands), so I won't be available for looking at this before a few weeks.

**Comment by lpw25** [comment]
> @lpw25 does that sound reasonable?

That matches my expectation too, although I'd like to see some exploration of where the cost is coming from, because I wouldn't be surprised in much of it could be avoided with a little effort.

**Comment by gasche** [comment]
If I understand correctly, both @Octachron and @lpw25 have skimmed this PR (and they seem to have ideas about the performance cost), but we don't have a full review for now.
My plan is to wait for one of them to do a review, or maybe enough time will have passed (maybe another month?) that I assume they are busy with other things, then I would try to find another volunteer or look it up myself.

**Comment by Drup** [comment]
I'm hugely in favour of doing changes to make code manipulating type expressions less easy to get wrong. So I had a look at the code. I apologize in advance, I'm going to be That Guy.

If we are improving the code "down there", I think we should take this occasion to also try to follow the absolute minimum in term of code quality. I'm going to suggest trivial general improvements for the *new* code that I just saw.
- Make the minimal documentation mildly usable (organized, valid ocamldoc, correctly attached to the right thing, etc)
- **Use labels for blind booleans arguments**
- Find understandable names for the operations (`newty3` `type_repr1` ???)
- Try to consolidate submodules to make sense (`TTypeSet` vs `TypeSet` ??)

While most of `btype`/`ctype`/... indeed does not follow these guidelines, the new code should.

For the patch itself: I think it goes in the right direction! I'm also not particularly worried about the perf aspect, but I think it can be improved in a later patch, since such a huge patch is going to cause many conflict if we let it linger (it already does, it seems)

**Comment by garrigue** [comment]
Tried to follow @Drup 's suggestions.
However, I could not find a better name for `newty3`: this is `Transient_expr.create`, omitting just the `id` argument.
And it is called only a few times.

As he very correctly says, this PR is basically in conflict with any change in the type checker, and I spent again one hour today rebasing it, so it would be nice if we could promptly merge.
By the way our analysis of the overhead does not show any single bottleneck, so we do not hope too much for a major speed-up in that direction.

**Comment by Octachron** [comment]
>By the way our analysis of the overhead does not show any single bottleneck, so we do not hope too much for a major speed-up in that direction.

Is the data available somewhere?

**Comment by t6s** [comment]
In the attached spreadsheet ([comparison.ods](https://github.com/ocaml/ocaml/files/6582976/comparison.ods)),
I compared three branches: trunk, always_repr and always_repr_transient_Subst,
and builds of four example programs: stdlib, ocamldoc, coqtop and lablgtk3.

The commands used for the builds are as follows.
stdlib, ocamldoc: `make clean; make`
coqtop: `dune build -j 1 --no-buffer dev/shim/coqtop-prelude`
lablgtk3: `dune build -p lablgtk3 -j 1 --no-buffer`

The new branch always_repr_transient_Subst contains an easy "optimization"
which changes `Ctype.copy` and `Subst.typexp` a bit to make them use
`Transient_expr` directly in order to reduce the number of calls to `repr` in these functions.


Sheet 1 lists measurements on the number of calls to `repr`

The values in columns titled trunk, always_repr and transient_Subst are raw counts of function calls.
Columns "Alw / trunk" and "TrSbst / trunk" show the factors of increases.
Columns "Alw - TrSbst" and "(H) / Alw" show how many calls are spared by the "optimization".

The rows titled stdlib, ocamldoc, coqtop and lablgtk3 are total numbers of calls in the corresponding builds.
Other rows, `Subst.typexp`, `unify` and `Ctype.copy` are the numbers of call stacks which contain
the corresponding functions.

For example, the `Subst.typexp` row just below stdlib indicates that there were 1068016 `repr`s
called directly or indirectly from `Subst.typexp` during a build of stdlib using trunk,
2775774 using always_repr, and so on.


Sheet 2 lists time measurements of the builds listed in Sheet 1.

The increases in the cases of stdlib and ocamldoc are rather large (6 to 9 percent).
Comparing them to the much smaller increase in coqtop (ca. 1.8 percent), Jacques and I suspect that
only the bytecode binaries of the compiler (ocamlc.byte, ocamlopt.byte) are significantly
affected, while the native versions are not so much.
(But I am not sure about what to say for lablgtk3, which shows an intermediate slowdown.)

The "optimization" in always_repr_transient_Subst branch is shown to be almost useless,
especially when using native versions.  It seems to be inferred that the major cause of the slowdown
is not the increased number of calls.


Raw CSVs are here:
 https://drive.google.com/drive/folders/1--47VGHsaOI5iDoA8c1-r4hzZAYS1bpp?usp=sharing

**Comment by gasche** [comment]
Thanks for the detailed analysis.

I think that a slowdown when using ocaml{c,opt}.byte (bytecode-compiled compilers) are not very important, because our users don't use those in production. I would focus on the slowdown measured with ocaml{c,opt}.opt (native-compiler compilers). Do I correctly understand that stdlib/ocamldoc use .byte compilers, and coqtop/lablgtk use .opt compilers?

Looking at the number of calls among Subst.typexp, unify and copy: it looks like your optimizations are very effective for Subst.typexp and copy, but the overall number of calls is dominated by unify. (I'd say the optimizations are still good to have if they keep the code simple/maintainable.)

Looking at the times: what did you measure exactly, just the type-checking time (using -dtimings / -dprofile) or overall compile time? If the latter, that may explain the different between coqtop and lablgtk: it may be that lablgtk spends a larger portion of type on the type-checking, so the same proportional increase in type-checking gives a larger overall increase.

**Comment by t6s** [comment]
The measurements of time are done for entire builds, i.e.,
by `make clean; time make` or `dune clean; time dune ...`.
Is there any good technique to profile just typecheking?

stdlib and ocamldoc are compiled simply issuing
`make clean; time make` in each subdirectory.
I am starting to doubt if it was actually using the
bytecode binaries of the compiler. Let me recheck it.

I agree that the optimizations are nice anyway to be included in this PR.
They are reverting the abstraction, but only in small portions.

**Comment by gasche** [comment]
> Is there any good technique to profile just typecheking?

For single modules, yes, passing `-dtimings` or `-dprofile` shows you a per-pass breakdown. For complete projects, we would need to aggregate those timings. (This could be done easily if those options respected `-dump-into-file`, then we could run a small script to aggregate the results after the build. But this is not easy to do right now.)

**Comment by t6s** [comment]
> stdlib and ocamldoc are compiled simply issuing
> `make clean; time make` in each subdirectory.
> I am starting to doubt if it was actually using the
> bytecode binaries of the compiler. Let me recheck it.

I have confirmed that stdlib was compiled using ocamlc.byte, but ocamldoc was using ocamlc.opt.
Coq and lablgtk were using ocamlopt.opt.

As @gasche suggested, the 2%, 3% and 6% losses at coq, lablgtk and ocamldoc seems to be corresponding to the
balance between typechecking and other steps of compilation.

**Comment by gasche** [comment]
My impression is that, with your performance analysis, the performance implications are well-understood and not too worrisome. We would need a code review for correctness to make progress.

**Comment by Octachron** [comment]
My understanding is more than the data is inconclusive and there is no explanation left to the slowdown. I will have a look at collecting the typechecking timings (probably by adding a `-dump-into-dir` option?).

**Comment by gasche** [comment]
Iirc., compiling `foo.ml` with `-dump-into-file` will generate a `foo.dump` file. This should be fine if we only pass `-dtimings` (and not some other dump flags). I think, for example, with dune we would have .dump copies of each .ml file in the build artifacts, and we can aggregate those. (Which requires writing a parser for the output format, which may require small tweaks to the format.)

File: typing/ctype.ml
Comment (gasche) [inline typing/ctype.ml:2715]: It looks like this fast path has been removed. Could this affect performance? In the new code it should still be possible to return early if `t1 == t2` and/or if `get_desc t1 == get_desc t2`.

**Comment by gasche** [comment]
My proposed explanation for the slowdown: type-checking on those examples is dominated by unification (and other non-repr-optimized parts of the type-checker), and traversing a huge number of small graphs with an extra function (non-inlinable) call on each node simply adds overhead, because the rest of the work is fairly small.

How to verify this: write a micro-benchmark that builds two large (identical) types and unifies them, in a loop many times. Measure overhead. Observe that the overhead is similar to the overhead on real-world programs (when looking only at the type-checkign time... which we are not sure how to measure).

File: typing/ctype.ml
Comment (gasche) [reply typing/ctype.ml:2715]: Actually there is a call to `eq_type t1 t2` at the beginning of `unify_eq` called below, so maybe it's fine. I note that `eq_type` checks `get_desct t1 == get_desc t2`, but not `t1 == t2` first, I wonder if this makes a difference.

Comment (garrigue) [reply typing/ctype.ml:2715]: The point is that, since normalization is done on access rather than on the node itself, the probability that two type nodes are physically equal gets lower (i.e., now normalization changes the `Tlink` inside the node, but does not change the node we are looking at, as it was the case with `repr`).
We could still add a fast path to `Types.eq_type`. This is an extra test, but this may avoid two calls to `repr` when successful.

**Comment by garrigue** [comment]
@gasche Do you mean that you want us to account for the cause of the precise overhead quantitatively?
It was clear from the start that there would be an overhead, since we do not pass around already normalized forms, and greatly increase the number of internal calls to `repr`.
More surprising are the result for Coq, i.e. what is cancelling the overhead, but considering the size of the codebase this is harder to analyze.

**Comment by garrigue** [comment]
We added the check, but this only reduces the calls to `repr` in `Types.eq_type` by 5% for stdlib, and in total by about 0.5% (`eq_type` causes about 10% of the calls to `repr`), so it has no measurable performance impact. We still keep the change as it should have no negative impact.

**Comment by gasche** [comment]
> @gasche Do you mean that you want us to account for the cause of the precise overhead quantitatively?

Personally I think it's fine, but @Octachron was asking for a better understanding of where the overhead comes from, or possibly what is the worst-case overhead on tyoe-checking-bound modules.

**Comment by Octachron** [comment]
I should mention that this is mostly to scratch my personal statistical data analysis itch for fun (and maybe understanding). Moreover, @gasche's hypothesis makes me wonder what happen on a flambda(and cross-module inlining)-enabled version of the compiler?

**Comment by gasche** [comment]
The `repr` function that we are talking about is a nasty recursive function with state update, I don't think it would be effectively inlined by an optimizer. (It has a short-path condition which would benefit from being hoisted out, but again this is probably too sophisticated for optimizers.)

**Comment by garrigue** [comment]
Actually, `Types.repr` is now split in 3 functions, with the first two non-recursive, so there is a potential for inlining.
Those non-recursive functions contain a pattern matching that requires at least two tests before reaching the default case (which does nothing).
I didn't look precisely how much inlining was done.

**Comment by garrigue** [comment]
This is getting problematic: we have again a bunch of conflicts so this will need to be rebased, but this time I will wait until somebody reviews.
We would like to base all our work on this new approach, but this is difficult until this PR is merged.
As a stopgap measure, we can create PRs against this branch.
For instance, there is now one making `row_desc` abstract: https://github.com/COCTI/ocaml/pull/1

**Review by Drup — APPROVED** [review]
I'm rather happy with the changes to my remarks (as always with code quality, we can do better, but at least I find the code easier to read now).

From my cursory view, the changes look correct (i.e., they achieve what it says on the PR). I do not have any hindsight on the perf aspect, I'll let @Octachron and/or @lpw25 comments on the finer details.

File: typing/ctype.ml
Comment (Octachron) [inline typing/ctype.ml]: This commented-out code can be removed.

**Comment by garrigue** [comment]
Thank you all!

**Comment by Octachron** [comment]
Note that I am still planning to do some measurement to both get a finer understanding of the performance regression and have a tool ready when similar issues arise again. I hope to have some preliminary results this week.

However, I agree that potential performance fixes can be implemented later, and simplifying the typechecker now is the right decision.

**Comment by Octachron** [comment]
It took more time than I expected (or did I went more overboard than I expected?), but I have finished a series of measure of compilation time using #10575 using  the opam packages for ocamlfind, num, zarith, seq, containers, coq, dune, re, ocamlbuild,uchar, topkg, uutf, tyxml, sexplib0, base. I have compiled each package with and without this PR 250 times, recording the compilation profile (time spent in the typechecker, and total compilation time) for each file in the package (the raw data is [here](https://www.polychoron.fr/static/longer_complex.log.xz)).

The average relative increase in typechecking time seems to be around 6%. The total compilation time only increases by 3%.
The quantiles also seem reasonable:

| quantile     | relative change of total compilation time |
|-------|-----------------|
| 1%    | 0.945555        |
| 10%   | 1               |
| 25%   | 1.00707         |
| 50%   | 1.02507         |
| 75%   | 1.05            |
| 90%   | 1.07895         |
| 99%   | 1.17846         |
| 99.9% | 1.4379          |

with more than 75% of the compiled files that see an increase of the total compilation time of less than 5%.

I have slightly more detailed analysis and many graphics at https://www.polychoron.fr/ocaml/2021/08/19/measuring_compilation_times.html (which is still a bit of a work in progress). The code used for the analysis can be found at  https://github.com/Octachron/ocaml-perfomance-monitoring .

**Comment by gasche** [comment]
Interesting! What does the 44%-increase code look like?

**Comment by t6s** [comment]
Great! Did you employ any technique to avoid thermal throttling and stabilize the CPU clock?

**Comment by Octachron** [comment]
@gasche : the worse increase is seen on:

- [base:src/applicative.mli](https://github.com/janestreet/base/blob/b7ca9b9/src/applicative_intf.ml): 17%
- dune:dune_rules__dir_status.mli: 18%
- [dune:dune_rules__merlin_ident.ml](https://github.com/ocaml/dune/blob/a181b93/src/dune_rules/merlin_ident.ml): 18%
- dune:dune_rules__env_node.mli: 20%
- [dune:dune_rules__compilation_context.mli](https://github.com/ocaml/dune/blob/66c846e/src/dune_rules/compilation_context.mli): 22%
- [dune:stdune__exn.ml](https://github.com/ocaml/dune/blob/66c846e655c6b677963c03cefafa35e08bf0a79e/otherlibs/stdune-unstable/exn.ml): 22%
- dune:dune_engine__rules.mli: 22%
- [coq:printing/printer.mli](https://github.com/coq/coq/blob/757d6ed/printing/printer.mli): 25%
- [dune:dune_engine__lib_name.mli](https://github.com/ocaml/dune/blob/497a4d6/src/dune_engine/lib_name.mli): 36%
- [dune:dune_rules__vimpl.mli](https://github.com/ocaml/dune/blob/497a4d6/src/dune_rules/vimpl.mli) : 44%

Some of those files are really small, so it might a be measurement error?
 
@t6s: I should have but it slipped my mind and I only took care of isolating the cpu doing the compilation. I was hoping that with enough data points, even the noise could be informative.

**Comment by gasche** [comment]
I edited your post to add links to (the current development version of) some of those files. Indeed I cannot discern a pattern, and they seem relatively small (but not one-liners either). Is there a way to reduce the noise of those small files? I can think of two ideas:
- simple: we could filter out the files whose type-checking time is below a certain threshold
- complex: we could do a first run to measure type-checking time, and then only compute actual statistics on a second run where small files have been aggregated together: files with below-threshold type-checking time (for the first run) have been grouped in "packs" (that go above the threshold), and only the collective runtime of each pack is reported

I think that the simplest strategy is good enough: in theory it's bad if you miss the fact that a lot of small files all got slower, but in practice I would expect large files to get slower as well in these scenarios, so the "simple" strategy should not miss the slowdown.

**Comment by Octachron** [comment]
Filtering files with small typechecking time is straightforward : this is already done to avoid NaN when computing the ratio of average time spent outside of the typechecker. I will also check another data set to see if the tail of the distribution is stable enough.

File: typing/btype.ml
Comment (trefis) [inline typing/btype.ml:650]: This looks like a bug. At the very least, it goes against the documentation of the function in the mli, which explicitly states "No [repr]'ing".

Comment (garrigue) [reply typing/btype.ml:650]: The comments in the mli were forgotten when switching to `always_repr`.
Repr'ing is now the correct behavior everywhere (except in `Printtyp`).
Fixed in #10665 .

**Pull request merged**
